usertable.cpp

```diff
@@ -250,6 +252,79 @@ UserTable::~UserTable()
   Dispose();
 }
 
+bool UserTable::IsDirectory(const std::string& sPath)
+{
+  static struct stat64 sStat;
+
+  if (-1 != lstat64(sPath.c_str(), &sStat) && S_ISDIR(sStat.st_mode)) {
+    return true;
+  }
+  else {
+    return false;
+  }
+}
+
+void UserTable::GetDirectoryTree(const std::string& sPath, TPathList& rResults)
+{
+  struct dirent* ent;
+  DIR* dir;
+
+  dir = opendir(sPath.c_str());
+  if (!dir) {
+    syslog(LOG_WARNING, "could not open directory %s: %s", sPath.c_str(), strerror(errno));
+    return;
+  }
+
+  rResults.push_back(sPath);
+
+  ent = readdir(dir);
+  while (ent) {
+    std::string sFullPath = sPath + "/" + ent->d_name;
+    std::string sEntry(ent->d_name);
+    if ("." != sEntry && ".." != sEntry && IsDirectory(sFullPath)) {
+      GetDirectoryTree(sFullPath, rResults);
+    }
+    ent = readdir(dir);
+  }
+  closedir(dir);
+}
+
+void UserTable::AddWatch(const std::string& sPath, IncronTabEntry* pEntry)
+{
+  InotifyWatch* pW = new InotifyWatch(sPath, pEntry->GetMask());
+
+  syslog(LOG_DEBUG, "adding directory watch %s", sPath.c_str());
+
+  // warning only - permissions may change later
+  if (!(m_fSysTable || MayAccess(sPath, DONT_FOLLOW(pEntry->GetMask()))))
+    syslog(LOG_WARNING, "access denied on %s - events will be discarded silently", sPath.c_str());
+
+  try {
+    m_in.Add(pW);
+    m_map.insert(IWCE_MAP::value_type(pW, pEntry));
+  } catch (InotifyException e) {
```

Comment:
Better to catch exceptions by reference for types that can be inherited.
Can save bad polymorphic casts.